SignalServiceKit/src/Storage/OWSPrimaryStorage.m

```diff
@@ -119,15 +124,72 @@ - (instancetype)initStorage
     if (self) {
         [self loadDatabase];
 
-        _dbReadConnection = self.newDatabaseConnection;
-        _dbReadWriteConnection = self.newDatabaseConnection;
+        _dbReadConnection = [self newDatabaseConnection];
+        _dbReadWriteConnection = [self newDatabaseConnection];
+        _uiDatabaseConnection = [self newDatabaseConnection];
+        
+        // Increase object cache limit. Default is 250.
+        _uiDatabaseConnection.objectCacheLimit = 500;
+        [_uiDatabaseConnection enableExceptionsForImplicitlyEndingLongLivedReadTransaction];
+        [_uiDatabaseConnection beginLongLivedReadTransaction];
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(yapDatabaseModified:)
+                                                     name:YapDatabaseModifiedNotification
+                                                   object:self.dbNotificationObject];
+
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(yapDatabaseModifiedExternally:)
+                                                     name:YapDatabaseModifiedExternallyNotification
+                                                   object:nil];
 
         OWSSingletonAssert();
     }
 
     return self;
 }
 
+- (void)yapDatabaseModifiedExternally:(NSNotification *)notification
+{
+    // Notify observers we're about to update the database connection
+    [[NSNotificationCenter defaultCenter] postNotificationName:OWSUIDatabaseConnectionWillUpdateExternallyNotification object:self.dbNotificationObject];
+    
+    // Move uiDatabaseConnection to the latest commit.
+    // Do so atomically, and fetch all the notifications for each commit we jump.
+    NSArray *notifications = [self.uiDatabaseConnection beginLongLivedReadTransaction];
```

Comment:
`beginLongLivedReadTransaction` is expensive.  When we send a message in the SAE, the main app will get 20-30 or more `YapDatabaseModifiedExternallyNotification`.  Previously, we weren't doing any work in response to them because our views were ignoring `YapDatabaseModifiedExternallyNotification` if the view wasn't visible and the app wasn't active.  

I'm not too concerned about the perf cost since the main app will be in the background in this case, but FYI.